DSS-3207: added memory setting usage parameter @ dss-pades-pdfbox

roy20021 opened the pull request:
Follow-up of https://github.com/esig/dss/pull/176

roy20021 commented:
@bsanchezb 

> Better to avoid changing the constructors of PdfBoxSignatureService, or other interfaces. It may have a big impact on the implementors. Instead, it would be better to configure the PDFSignatureService with a setter, as done in [AbstractPdfObjFactory#configure](https://github.com/esig/dss/blob/6.1.x/dss-pades/src/main/java/eu/europa/esig/dss/pdf/AbstractPdfObjFactory.java#L97).

Done. Actually, I have changed some PDFBoxUtil's methods' interface.

> As you mentioned, support of similar setting in OpenPdf implementation would be highly desirable. Could you please evaluate the possibility to implement it? The [MemoryUsageSetting](https://github.com/aclarkxyz/PDFBoxLite/blob/master/src/org/apache/pdfbox/io/MemoryUsageSetting.java) from PdfBox looks to be easy to configure, so maybe we could introduce our own DSS wrapper for memory settings to be shared between two implementations, and to be transformed to implementation specific class when needed?

Done, however the mapping cannot be 1-to-1 due of different level of underlying support

> For the formatting, try to simply follow our style. Please add JavaDocs for any protected or public methods and classes, if possible. Simply add a small description on what method does and comments for input and output objects. See other classes as an example.

I tried.. I propose (and I can do in another ad-hoc PR) to configure spotless to also format the code

> For the testing you may see what we do in unit test [PAdESLevelBSignWithTempFileHandlerTest.java](https://github.com/esig/dss/blob/6.1.x/dss-pades/src/test/java/eu/europa/esig/dss/pades/signature/suite/PAdESLevelBSignWithTempFileHandlerTest.java), retrieving the memory usage in runtime (not ideal, but works well for big memory differences).

Done but I am not satisfied of the results. Often the memory consumption isn't reliable and test may fail. 

Anyway the activity is fine, for example If I run test with small max heap value the memory only setting test fails instead the file one succeed.
Moreover, I think that the unit test should be about using properly the underlying libraries.. not checking their behaviour so unit testing also them.
Can you have a look here?

roy20021 commented:
@bsanchezb can you have a look, please?

Maybe there are too changes.. do I need to remove the license commit?

roy20021 commented:
> Dear Andrea,
> 
> Thank you for the PR, it looks very good.
> 
> I have added a few comments directly in the code, could you please take a look and address the issues?
> 
> I would also highly appreciate, if you could add a documentation about the near future, within the cookbook, just after the [DSSResourcesHandler](https://ec.europa.eu/digital-building-blocks/DSS/webapp-demo/doc/dss-documentation.html#DSSResourcesHandler) chapter.
> 
> KR, Aleksandr

@bsanchezb I do not see any comments, have you submitted a review?

bsanchezb reviewed (changes requested):
Please start from using PdfMemoryUsageSetting directly in PdfBoxDocumentReader/ITextDocumentReader constructors instead of implementation-specific classes. This should simplify the code and leave conversion to the very end.

bsanchezb commented:
Hello @roy20021,

Thank you for the comment. I did not know it was not visible to you. Now you should have the review available.
Should you have any question, please do not hesitate to ask.

KR,
Aleksandr

roy20021 commented on dss-pades-openpdf/src/main/java/eu/europa/esig/dss/pdf/openpdf/ITextDocumentReader.java:
```diff
@@ -102,13 +107,25 @@ public ITextDocumentReader(DSSDocument dssDocument) throws IOException, InvalidP
+				if (ITextPdfMemoryUsageSetting.Mode.FULL.equals(memoryUsageSetting.getMode())) {
+					this.pdfReader = new PdfReader(filenameSource, passwordProtection);
+				} else {
+					this.pdfReader = new PdfReader(new RandomAccessFileOrArray(filenameSource), passwordProtection);
```
I reply starting from the last sentences.

There are 2 different but very related topics: loading mode (full or partial) and memory setting (file, memory).
Clearly, they are related because if the loading mode is full than all pdf has to be loaded in memory.

At best of my knowledge, each library offers different level of configuration among such topics.
For OpenPdf, as a conversational reference, take a look at: https://stackoverflow.com/questions/62164906/use-of-randomaccessfileorarray
Check by yourself the code of the PdfReader constructors regards the last loading line: _readPdfPartial();_ for RandomAccessFileOrArray or _readPdf();_ for the other ones.

PdfBox models it differently and inherently uses partial mode when file usage is requested.

----

Regards `new RandomAccessFileOrArray(filenameSource, true, false)` won't be okay because will load all pdf into memory so as a full load (instead of partial load).
Anyway, thanks to your suggestion, I notice the _com.lowagie.text.Document.plainRandomAccess_ attribute and the same semantic into the RandomAccessFileOrArray constructor.
In case of FileDocument AND partial loading I can set that flag to true so the content will be loaded without mapping memory.

I see here an issue of expressivity of the common model and the implementation specific ones. It might be desirable to config whatever combination but I guess that wrapping it is not a solution so just 'reasonable' combination can be provided.

bsanchezb replied:
Ok, you can leave it for now. I will take a look later in more detail

roy20021 commented on dss-pades-openpdf/src/main/java/eu/europa/esig/dss/pdf/openpdf/ITextPdfMemoryUsageSetting.java:
```diff
@@ -0,0 +1,41 @@
+	 * Load options
+	 */
+	public enum Mode {
+		FULL, PARTIAL
```
I have commented elsewhere regards FULL/PARTIAL and FILE/MEMORY topics. Hope that clarify.

roy20021 commented on dss-pades/src/test/java/eu/europa/esig/dss/pades/signature/suite/PAdESLevelBSignWithFileMemoryUsageSettingTest.java:
```diff
@@ -0,0 +1,227 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.tuple.Pair;
```
There is commented code that I left waiting your advice that's because...

Please have a look at the other comment about 'unit testing' the pdf libraries.

roy20021 commented on dss-pades-pdfbox/src/test/java/eu/europa/esig/dss/pades/PdfBoxUtilsTest.java:
```diff
@@ -60,42 +63,42 @@ void init() {
 	@Test
 	void generateScreenshotTest() {
-		DSSDocument screenshot = PdfBoxUtils.generateScreenshot(sampleDocument, 1);
+		DSSDocument screenshot = PdfBoxUtils.generateScreenshot(sampleDocument, 1, MemoryUsageSetting.setupMainMemoryOnly());
```
do you want I create both version of each method?
I won't keep only the old PdfBoxUtils methods since that will always load into memory and it is not the deal. Am I correct?

bsanchezb replied:
Yes, it would be desirable to have overloaded methods, so simple as well as advanced configuration is possible.
If it becomes too many of variables, we may also think about creation of a separate class for screenshot generation, for instance using a builder or a factory design patterns.

roy20021 commented on dss-pades/src/main/java/eu/europa/esig/dss/pdf/PdfMemoryUsageSetting.java:
```diff
@@ -0,0 +1,114 @@
+	 * MemoryUsageSetting's load options
+	 */
+	public enum Mode {
+		MEMORY, FILE, MIXED;
```
As per other comments, the mapping is pretty implementation biased. Shall I write that here?

roy20021 commented:
@bsanchezb I have commented your observations, can you see them?

bsanchezb commented:
Hi Andrea,

Yes, I can see them, thank you. I will take a look on them later, currently I do not have much time.

Best regards,
Aleksandr

bsanchezb commented on dss-pades-pdfbox/src/main/java/eu/europa/esig/dss/pdf/pdfbox/PdfBoxDocumentReader.java:
```diff
@@ -104,18 +104,24 @@ public PdfBoxDocumentReader(DSSDocument dssDocument)
+			if (dssDocument instanceof FileDocument) {
 				FileDocument fileDocument = (FileDocument) dssDocument;
 				this.pdDocument = PDDocument.load(fileDocument.getFile(), passwordProtection, memoryUsageSetting);
+			} else if (dssDocument instanceof InMemoryDocument) {
```
Ok

bsanchezb commented on dss-pades/src/test/resources/big_file_26mb.pdf:
> I think it is the wrong approach to verify the memory consumption. I propose to check that the PDF libs are called as supposed to be and not "unit testing" also their behaviour.

Yes, we may follow this approach instead

bsanchezb commented on dss-pades-pdfbox/src/main/java/eu/europa/esig/dss/pdf/pdfbox/PdfBoxDocumentReader.java:
```diff
@@ -104,18 +104,24 @@ public PdfBoxDocumentReader(DSSDocument dssDocument)
 	 */
-	public PdfBoxDocumentReader(DSSDocument dssDocument, String passwordProtection, MemoryUsageSetting memoryUsageSetting)
+	public PdfBoxDocumentReader(DSSDocument dssDocument, String passwordProtection,
+			MemoryUsageSetting memoryUsageSetting)
```
> Instead, it will not avoid the creation of the specific OpenPdf one.

Should be possible to avoid, but you may start the implementation and we will see after how we can adjust the OpenPdf  implementation

bsanchezb commented:
Hi @roy20021 ,

I did a quick review of your proposals. In general I agree with your arguments. You may try to implement the proposals with constructors and methods arrangement and we will see if it has any additional room for improvement.
We may also adapt the unit tests as you proposed.

KR,
Aleksandr

roy20021 commented:
@bsanchezb I implemented your proposal around constructors and method re-arrangement. I think it's pretty concise now.
I have added the scenario in openpdf when is FILE and partial loading.

Can you re-work the unit test with the agreed approach?

About doc, I haven't touched it yet. Waiting for feedback in order to have a stable solution to document.

When will be the next release? Do you think we can make this to be in?

Regards,
Andrea

roy20021 commented:
Hello @bsanchezb , 

Today I was trying to run spotless on a new branch starting from dss/develop in order to reduce the diss in this PR. However, the build is broken. Tests are failing.
Can you have a look, please?

About this point, any reasons why no CI is in place with the github actions? It would enforce the process, I see it high valuable for maintenance.

bsanchezb commented:
Thank you @roy20021 .

I will continue on this after the upgrade to PdfBox 3 (see [DSS-3122](https://ec.europa.eu/digital-building-blocks/tracker/browse/DSS-3122)). Hopefully, both will be included in DSS 6.2.

Best regards,
Aleksandr

roy20021 commented:
Hey @bsanchezb , 

I have collected some statistics. Here the results: [DSS3207 - Comparisons.xlsx](https://github.com/user-attachments/files/17558950/DSS3207.-.Comparisons.xlsx)

My observations:
1. the PR works as expected. The bigger the file is, clearer is the trend. I have also tried to set "-Xmx" with small values and memory approach fails OOM instead file one succeed.
2. openpdf performs pretty bad compared to pdfbox. It crashes with the 360 MB file
3. something "expensive" is happing around the T/LT/LTA signing... either memory or file approach requires much more memory. I haven't digged into it, is it a well known behaviour intrinsic of such signature types?
Sometimes, the trend might be even inverted (memory cheaper than file).

I think I did everything I could for contributing here. I let you continue on this.

Please let me know about point (3), I would like to understand more about it since it requires so much memory.

As a future improvement over the same topic, I believe it's worthful to handle the signature validations with a more lightly mode (again streaming, temp file, other).

Regards,
Andrea

bsanchezb commented:
Hi @roy20021 ,

Thank you for the latest changes and insightful statistics.

I have continued the work on the PR from a PdfBox 3 branch. You may find my take on this in PR #179.

The following has been done:

1. The branch has been merged with develop and PdfBox3 branches.
2. ITextPdfMemoryUsageSetting has been removed, as no longer needed (as it is the only purpose was to convert PdfMemoryUsageSetting.Mode.MEMORY into ITextPdfMemoryUsageSetting.Mode.FULL);
3. Added a special handling for InMemoryDocument in ITextDocumentReader, similar to the one you introduced in PdfBoxDocumentReader (good catch by the way);
4. Simplified the PAdESLevelBSignWithFileMemoryUsageSettingTest class, to reduce number of computations (as the test is mainly for a manual validation), and to get rid of Apache Commons dependency (please note that this blocked execution of the test with dss-utils-google-guava module). I think the dedicated objects are also clearer than working with Pair's. Additionally, I have added a similar memory validation on #validateDocument method.
5. Probably the biggest change I did, is the introduction of PdfMemoryUsageSetting.Mode.MEMORY_FULL enumeration, allowing to get back to the previous behavior with a FileDocument being completely loaded to binaries before providing the file to a PdfDocumentReader. From my testing, I have noticed the difference between PdfBox's and OpenPdf's "in-memory" handling with a File parameter provided in a constructor, against the byte[] array (or InputStream) provided directly. With MEMORY_FULL approach behaving usually better performance-wise. You may see the results of my tests below. Considering the major difference in OpenPdf implementation, I decided to keep the MEMORY_FULL mode to be used by default, so users won't get an unexpected performance downgrade after the new release. Please note that the old Mode.MEMORY has been renamed to a Mode.MEMORY_BUFFERED. I hope it is logically correct. 
6. PdfBoxUtils methods have been restored and deprecated in favor of a new class PdfBoxScreenshotBuilder. I think it is a good compromise to your approach with a PdfBoxDocumentRequest class, which I could not accept as we usually do not use this design pattern, and additionally I wanted to provide a smooth migration to the end-user.
7. Added documentation and migration guide.

I'm looking forward to receiving a feedback from you, as I hope I did not brake the logic you implemented anywhere :)

> something "expensive" is happing around the T/LT/LTA signing... either memory or file approach requires much more memory. I haven't digged into it, is it a well known behaviour intrinsic of such signature types?
Sometimes, the trend might be even inverted (memory cheaper than file).

On signature extension, DSS extracts signatures from the provided document and performs their validation in order to ensure their validity before proceeding with augmentation. This may have some impact on the performance and memory consumption. Maybe it would be possible to improve with only one instance of a PdfDocumentReader being created, but this will require a significant refactoring across a few parts of the code. We will see if may get back to this issue later.

Below you may find signature creation, augmentation and validation metrics with the PR:

PdfBox execution tests:
```
+-----------------+-----------------+--------+--------------+
|                 | Step Name       | RAM    | Time spent   +
+-----------------+-----------------+--------+--------------+
| Memory Full     | getDataToSign   | 0016.0 | PT0.4248402S |
| Memory Buffered | getDataToSign   | 0016.0 | PT0.697874S  |
| File            | getDataToSign   | 0015.0 | PT0.649876S  |
| Memory Full     | tokenSign       | 0000.0 | PT0.0025074S |
| Memory Buffered | tokenSign       | 0000.0 | PT0.0038823S |
| File            | tokenSign       | 0000.0 | PT0.0020271S |
| Memory Full     | signDocument    | 0002.4 | PT0.6436295S |
| Memory Buffered | signDocument    | 0002.4 | PT0.6671888S |
| File            | signDocument    | 0001.9 | PT0.6409686S |
| Memory Full     | T-Level extendD | 0019.3 | PT3.7269212S |
| Memory Buffered | T-Level extendD | 0016.1 | PT3.8317178S |
| File            | T-Level extendD | 0041.7 | PT3.7065721S |
| Memory Full     | LT-Level extend | 0030.8 | PT4.3653148S |
| Memory Buffered | LT-Level extend | 0019.4 | PT4.4845114S |
| File            | LT-Level extend | 0017.5 | PT4.6116488S |
| Memory Full     | LTA-Level exten | 0058.3 | PT5.7084445S |
| Memory Buffered | LTA-Level exten | 0034.9 | PT6.0610176S |
| File            | LTA-Level exten | 0034.8 | PT6.8180666S |
+-----------------+-----------------+--------+--------------+

+-----------------+-----------------+--------+--------------+
|                 | Step Name       | RAM    | Time spent   +
+-----------------+-----------------+--------+--------------+
| Memory Full     | validateDocumen | 0090.5 | PT1.3167352S |
| Memory Buffered | validateDocumen | 0075.2 | PT1.4526008S |
| File            | validateDocumen | 0062.5 | PT1.4617258S |
+-----------------+-----------------+--------+--------------+
```

OpenPdf execution tests:
```
+-----------------+-----------------+--------+--------------+
|                 | Step Name       | RAM    | Time spent   +
+-----------------+-----------------+--------+--------------+
| Memory Full     | getDataToSign   | 0067.0 | PT0.8610524S |
| Memory Buffered | getDataToSign   | 0038.0 | PT0.8403175S |
| File            | getDataToSign   | 0026.0 | PT0.8782862S |
| Memory Full     | tokenSign       | 0000.0 | PT0.0080129S |
| Memory Buffered | tokenSign       | 0000.1 | PT0.0039421S |
| File            | tokenSign       | 0000.0 | PT0.0029967S |
| Memory Full     | signDocument    | 0002.0 | PT0.7147587S |
| Memory Buffered | signDocument    | 0002.0 | PT0.7012383S |
| File            | signDocument    | 0002.0 | PT0.6749436S |
| Memory Full     | T-Level extendD | 0085.8 | PT3.9135983S |
| Memory Buffered | T-Level extendD | 0031.8 | PT3.8990954S |
| File            | T-Level extendD | 0034.2 | PT53.4609146 |
| Memory Full     | LT-Level extend | 0235.7 | PT4.2406194S |
| Memory Buffered | LT-Level extend | 0278.9 | PT4.2919467S |
| File            | LT-Level extend | 0042.1 | PT1M17.81157 |
| Memory Full     | LTA-Level exten | 0179.0 | PT5.4025485S |
| Memory Buffered | LTA-Level exten | 0090.9 | PT7.2816534S |
| File            | LTA-Level exten | 0082.2 | PT2M3.210222 |
+-----------------+-----------------+--------+--------------+

+-----------------+-----------------+--------+--------------+
|                 | Step Name       | RAM    | Time spent   +
+-----------------+-----------------+--------+--------------+
| Memory Full     | validateDocumen | 0328.2 | PT4.1754672S |
| Memory Buffered | validateDocumen | 0160.6 | PT4M13.54078 |
| File            | validateDocumen | 0089.8 | PT4M33.59037 |
+-----------------+-----------------+--------+--------------+
```
